### PR TITLE
Path to multiGPU: Basic utilities for cuISTL

### PR DESCRIPTION
**Pull request by kjetilly**
We are creating a new linear solver GPU path compatible with the DUNE ISTL library. This first PR provides basic utilities for handling CUDA, CUSPARSE and CUBLAS resources, as well as safe call macros to ensure safe programming (checks for return codes and throws an exception with a meaningful error message).

[This is related to the issue 3932](https://github.com/OPM/opm-simulators/issues/3932)

[This PR partially replaces PR 4137](https://github.com/OPM/opm-simulators/pull/4137)

This PR is a prerequisite for [PR 4634](https://github.com/OPM/opm-simulators/pull/4634), [PR 4635](https://github.com/OPM/opm-simulators/pull/4635) and [PR4636](https://github.com/OPM/opm-simulators/pull/4636)

### CODE REVIEW

**Comment by kjetilly** [comment]
So it seems the test fails because Jenkins does not have a GPU installed, but does have CUDA enabled. What is the best way to deal with this? Ideally, the tests in `tests/cuistl` should only be run when a CUDA compatible GPU is present. Is there any change I need to do to the code to accommodate this?

File: opm/simulators/linalg/cuistl/detail/cusparse_safe_call.hpp [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+#undef CHECK_CUSPARSE_ERROR_TYPE
+
+/**
+ * @brief cusparseSafeCall checks the return type of the CUSPARSE expression (function call) and throws an exception if
+ * it does not equal CUSPARSE_STATUS_SUCCESS.
+ *
+ * Example usage:
+ * @code{.cpp}
+ * #include <opm/simulators/linalg/cuistl/detail/cuda_safe_call.hpp>
```
Comment (atgeirr) [inline opm/simulators/linalg/cuistl/detail/cusparse_safe_call.hpp:64]: Example refers to cudaSafeCall here and below.

File: opm/simulators/linalg/cuistl/detail/cuda_check_last_error.hpp [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+ *
+ * void some_function() {
+ *     OPM_CUDA_CHECK_LAST_ERROR;
+ * }
+ * @endcode
+ *
+ * @note This can be used to debug the code, or simply make sure that no error has occured.
+ */
+#define OPM_CUDA_CHECK_LAST_ERROR OPM_CUDA_SAFE_CALL(cudaGetLastError())
+#ifdef NDEBUG
```
Comment (atgeirr) [inline opm/simulators/linalg/cuistl/detail/cuda_check_last_error.hpp:81]: Add a blank line above here, to separate the macros, and move the doc from below to outside the #ifdef NDEBUG.

File: opm/simulators/linalg/cuistl/detail/cuda_check_last_error.hpp [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+#define OPM_CUDA_CHECK_LAST_ERROR_IF_DEBUG
+#else
+/**
+ * @brief OPM_CUDA_CHECK_LAST_ERROR_IF_DEBUG checks the return type of cudaGetLastError() only if NDEBUG is not defined,
+ * and throws an exception if cudaGetLastError() does not equal cudaSuccess.
+ *
+ * Example usage:
+ * @code{.cpp}
+ * #include <opm/simulators/linalg/cuistl/detail/cuda_safe_call.hpp>
```
Comment (atgeirr) [inline opm/simulators/linalg/cuistl/detail/cuda_check_last_error.hpp:90]: Wrong include, wrong macro down below.

File: opm/simulators/linalg/cuistl/detail/CuBlasHandle.cpp [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+
+CuBlasHandle::CuBlasHandle()
+{
+    OPM_CUBLAS_SAFE_CALL(cublasCreate(&m_handle));
+}
+
+CuBlasHandle::~CuBlasHandle()
+{
+    OPM_CUBLAS_SAFE_CALL(cublasDestroy(m_handle));
```
Comment (atgeirr) [inline opm/simulators/linalg/cuistl/detail/CuBlasHandle.cpp:33]: Destructors should never throw. It is therefore better to not wrap this in OPM_CUBLAS_SAFE_CALL. Exceptions in destruction of static object is not really possible to handle anyway.

I think the only error possible from cublasDestroy() is CUBLAS_STATUS_NOT_INITIALIZED, which should be impossible (?) so it is not necessary anyway.

File: opm/simulators/linalg/cuistl/detail/CuSparseHandle.hpp [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+     * @brief getInstance creates (if necessary) and returns the single unique instance of CuSparseHandle (singleton)
+     */
+    static CuSparseHandle& getInstance();
+
+private:
+    CuSparseHandle();
+    cusparseHandle_t m_handle;
+};
+
+using CuSparseHandlePtr = std::unique_ptr<CuSparseHandle>;
```
Comment (atgeirr) [inline opm/simulators/linalg/cuistl/detail/CuSparseHandle.hpp:67]: Is this used anywhere?

**Review by atgeirr — APPROVED** [review]
Overall, I think this looks good! It may merged when all minor issues commented on have been addressed, and the building and testing is green on Jenkins. (The last point probably requires changes that I am not an expert on.)

**Comment by akva2** [comment]
https://github.com/OPM/opm-common/pull/3514 plus adding labels as in https://github.com/OPM/opm-simulators/pull/4637 will allow you to avoid the test failures.

**Review by atgeirr — APPROVED** [review]

**Comment by atgeirr** [comment]
jenkins build this please

**Comment by atgeirr** [comment]
jenkins build this please

**Comment by atgeirr** [comment]
All green, merging!

**Pull request merged**